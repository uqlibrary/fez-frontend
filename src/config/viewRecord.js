import { pathConfig } from 'config';

const prefixFileNameSuffix = (prefix = '', fileName, suffix = '', extension) =>
    `${prefix}${prefix !== '' ? '_' : ''}${fileName.substr(0, fileName.lastIndexOf('.'))}${
        suffix !== '' ? '_' : ''
    }${suffix}.${extension}`;

// const prefixFileNameSuffix = (prefix, suffix, fileName, extension) =>
//     `${prefix}_${fileName.substr(0, fileName.lastIndexOf('.'))}_${suffix}.${extension}`;

export const viewRecordsConfig = {
    genericDataEmail: 'data@library.uq.edu.au',
    licenseLinks: {
        453608: {
            className: 'cc-by',
            url: 'http://creativecommons.org/licenses/by/3.0/deed.en_US',
        },
        453609: {
            className: 'cc-by-nd',
            url: 'http://creativecommons.org/licenses/by-nd/3.0/deed.en_US',
        },
        453610: {
            className: 'cc-by-nc',
            url: 'http://creativecommons.org/licenses/by-nc/3.0/deed.en_US',
        },
        453611: {
            className: 'cc-by-nc-nd',
            url: 'http://creativecommons.org/licenses/by-nc-nd/3.0/deed.en_US',
        },
        453612: {
            className: 'cc-by-nc-sa',
            url: 'http://creativecommons.org/licenses/by-nc-sa/3.0/deed.en_US',
        },
        453613: {
            className: 'cc-by-sa',
            url: 'http://creativecommons.org/licenses/by-sa/3.0/deed.en_US',
        },
        456710: {
            className: 'cc-by',
            url: 'http://creativecommons.org/licenses/by/4.0/deed.en',
        },
        456711: {
            className: 'cc-by-sa',
            url: 'http://creativecommons.org/licenses/by-sa/4.0/deed.en',
        },
        456712: {
            className: 'cc-by-nd',
            url: 'http://creativecommons.org/licenses/by-nd/4.0/deed.en',
        },
        456713: {
            className: 'cc-by-nc',
            url: 'http://creativecommons.org/licenses/by-nc/4.0/deed.en',
        },
        456714: {
            className: 'cc-by-nc-sa',
            url: 'http://creativecommons.org/licenses/by-nc-sa/4.0/deed.en',
        },
        456715: {
            className: 'cc-by-nc-nd',
            url: 'http://creativecommons.org/licenses/by-nc-nd/4.0/deed.en',
        },
        // Permitted Re-use with Acknowledgement
        453701: {
            className: 'uq',
            url: 'https://guides.library.uq.edu.au/deposit-your-data/license-reuse-with-acknowledgement',
        },
        // Permitted Non-commercial Re-use with Acknowledgement
        453702: {
            className: 'uq',
            url: 'https://guides.library.uq.edu.au/deposit_your_data/terms_and_conditions',
        },
        // Permitted Re-Use with Commercial Use Restriction
        454104: {
            className: 'uq',
            url: 'https://guides.library.uq.edu.au/deposit-your-data/license-reuse-noncommercial',
        },
        // Permitted Re-Use with Share Alike Requirement
        454105: {
            className: 'uq',
            url: 'https://guides.library.uq.edu.au/deposit-your-data/license-reuse-share-alike',
        },
        // Permitted reuse only with a Data Sharing Agreement in place between UQ and recipient
        456807: {
            className: 'uq',
            url: 'https://guides.library.uq.edu.au/deposit-your-data/license-reuse-data-agreement',
        },
    },
    htmlFields: [
        'rek_transcript',
        'rek_notes',
        'rek_project_description',
        'rek_acknowledgements',
        'rek_advisory_statement',
        'rek_related_datasets',
        'rek_related_publications',
    ],
    // apart from rek_date
    dateFields: [
        'rek_date_available',
        'rek_date_recorded',
        'rek_date_photo_taken',
        'rek_date_scanned',
        'rek_start_date',
        'rek_end_date',
        'rek_time_period_start_date',
        'rek_time_period_end_date',
        'rek_project_start_date',
    ],
    dateFieldFormat: {
        rek_date_available: 'YYYY',
        rek_date_photo_taken: 'YYYY',
    },
    // some display types have different publication date format
    publicationDateFormat: {
        Book: 'YYYY',
        'Book Chapter': 'YYYY',
        'Conference Paper': 'YYYY',
        'Data Collection': 'YYYY',
    },
    files: {
        blacklist: {
            namePrefixRegex: '^(FezACML|stream|web_|thumbnail_|preview_|presmd)',
            nameSuffixRegex: '(_t\\.)',
            collections: ['UQ:413806', 'UQ:357493', 'UQ:211157', 'UQ:342107'],
        },
        thumbnailFileName: fileName => [
            prefixFileNameSuffix('thumbnail', fileName, 'compressed_t', 'jpg'),
            prefixFileNameSuffix('thumbnail', fileName, 'override_t', 'jpg'),
            prefixFileNameSuffix('thumbnail', fileName, 't', 'jpg'),
            prefixFileNameSuffix('thumbnail', fileName, '', 'jpg'),
        ],
        previewFileName: fileName => [
            prefixFileNameSuffix('preview', fileName, 'compressed_t', 'jpg'),
            prefixFileNameSuffix('preview', fileName, 'override_t', 'jpg'),
            prefixFileNameSuffix('preview', fileName, 't', 'jpg'),
            prefixFileNameSuffix('preview', fileName, '', 'jpg'),
        ],
        webFileName: fileName => [
            prefixFileNameSuffix('web', fileName, 'compressed_t', 'jpg'),
            prefixFileNameSuffix('web', fileName, 'override_t', 'jpg'),
            prefixFileNameSuffix('web', fileName, 't', 'jpg'),
            prefixFileNameSuffix('web', fileName, '', 'jpg'),
        ],
    },
    metaTags: [
        {
            searchKey: null,
            subkey: 'rek_pid',
            tags: [
                {
                    name: 'DC.Identifier',
                    isMultiple: false,
                },
            ],
            url: pid => pathConfig.records.view(pid, true),
        },
        {
            searchKey: 'fez_record_search_key_subject',
            subkey: 'rek_subject',
            tags: [
                {
                    name: 'DC.Subject',
                    isMultiple: true,
                },
            ],
        },
        {
            searchKey: null,
            subkey: 'fez_datastream_info',
            tags: [
                {
                    name: 'citation_pdf_url',
                    isMultiple: true,
                },
            ],
            // TODO add dsi_checksum as a third param, if this function ever gets used in the future
            url: (pid, fileName) => pathConfig.file.url(pid, fileName),
        },
        {
            searchKey: null,
            subkey: 'rek_title',
            tags: [
                {
                    name: 'DC.Title',
                    isMultiple: false,
                },
                {
                    name: 'citation_title',
                    isMultiple: false,
                },
            ],
        },
        {
            searchKey: 'fez_record_search_key_contributor',
            subkey: 'rek_contributor',
            tags: [
                {
                    name: 'DC.Contributor',
                    isMultiple: true,
                },
            ],
        },
        {
            searchKey: 'fez_record_search_key_author',
            subkey: 'rek_author',
            tags: [
                {
                    name: 'DC.Creator',
                    isMultiple: true,
                },
                {
                    name: 'citation_authors',
                    isMultiple: false,
                },
            ],
        },
        {
            searchKey: 'fez_record_search_key_journal_name',
            subkey: 'rek_journal_name',
            tags: [
                {
                    name: 'citation_journal_title',
                    isMultiple: false,
                },
            ],
        },
        {
            searchKey: 'fez_record_search_key_issn',
            subkey: 'rek_issn',
            tags: [
                {
                    name: 'citation_issn',
                    isMultiple: true,
                },
            ],
        },
        {
            searchKey: null,
            subkey: 'rek_date',
            tags: [
                {
                    name: 'DC.Date',
                    isMultiple: false,
                    format: 'YYYY-MM-DD',
                },
                {
                    name: 'citation_date',
                    isMultiple: false,
                    format: 'YYYY/MM/DD',
                },
            ],
        },
        {
            searchKey: 'fez_record_search_key_start_date',
            subkey: 'rek_start_date',
            tags: [
                {
                    name: 'citation_start_date',
                    isMultiple: false,
                    format: 'YYYY/MM/DD',
                },
            ],
        },
        {
            searchKey: 'fez_record_search_key_end_date',
            subkey: 'rek_end_date',
            tags: [
                {
                    name: 'citation_end_date',
                    isMultiple: false,
                    format: 'YYYY/MM/DD',
                },
            ],
        },
        {
            searchKey: 'fez_record_search_key_doi',
            subkey: 'rek_doi',
            tags: [
                {
                    name: 'citation_doi',
                    isMultiple: false,
                },
            ],
        },
        {
            searchKey: 'fez_record_search_key_volume_number',
            subkey: 'rek_volume_number',
            tags: [
                {
                    name: 'citation_volume',
                    isMultiple: false,
                },
            ],
        },
        {
            searchKey: 'fez_record_search_key_issue_number',
            subkey: 'rek_issue_number',
            tags: [
                {
                    name: 'citation_issue',
                    isMultiple: false,
                },
            ],
        },
        {
            searchKey: 'fez_record_search_key_start_page',
            subkey: 'rek_start_page',
            tags: [
                {
                    name: 'citation_firstpage',
                    isMultiple: false,
                },
            ],
        },
        {
            searchKey: 'fez_record_search_key_end_page',
            subkey: 'rek_end_page',
            tags: [
                {
                    name: 'citation_lastpage',
                    isMultiple: false,
                },
            ],
        },
        {
            searchKey: 'fez_record_search_key_publisher',
            subkey: 'rek_publisher',
            tags: [
                {
                    name: 'DC.Publisher',
                    isMultiple: false,
                },
                {
                    name: 'citation_publisher',
                    isMultiple: false,
                },
            ],
        },
        {
            searchKey: 'fez_record_search_key_language',
            subkey: 'rek_language',
            tags: [
                {
                    name: 'citation_language',
                    isMultiple: true,
                },
            ],
        },
        {
            searchKey: null,
            subkey: 'rek_description',
            tags: [
                {
                    name: 'DC.Description',
                    isMultiple: false,
                },
                {
                    name: 'citation_abstract',
                    isMultiple: false,
                },
            ],
        },
        {
            searchKey: 'fez_record_search_key_keywords',
            subkey: 'rek_keywords',
            tags: [
                {
                    name: 'citation_keywords',
                    isMultiple: false,
                },
            ],
        },
        {
            searchKey: 'fez_record_search_key_conference_name',
            subkey: 'rek_conference_name',
            tags: [
                {
                    name: 'citation_conference',
                    isMultiple: false,
                },
            ],
        },
        {
            searchKey: 'fez_record_search_key_coverage_period',
            subkey: 'rek_coverage_period',
            tags: [
                {
                    name: 'DC.Subject',
                    isMultiple: true,
                },
            ],
        },
        {
            searchKey: 'fez_record_search_key_geographic_area',
            subkey: 'rek_geographic_area',
            tags: [
                {
                    name: 'DC.Subject',
                    isMultiple: true,
                },
            ],
        },
        {
            searchKey: 'fez_record_search_key_isbn',
            subkey: 'rek_isbn',
            tags: [
                {
                    name: 'citation_isbn',
                    isMultiple: true,
                },
            ],
        },
        {
            searchKey: 'fez_record_search_key_report_number',
            subkey: 'rek_report_number',
            tags: [
                {
                    name: 'citation_technical_report_number',
                    isMultiple: false,
                },
            ],
        },
    ],
};

export const STATE_ADDED = 'A';
export const STATE_DELETED = 'D';
