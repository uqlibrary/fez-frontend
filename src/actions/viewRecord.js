import * as actions from './actionTypes';
import { get } from 'repositories/generic';
import { EXISTING_RECORD_API, EXISTING_RECORD_HISTORY_API, EXISTING_RECORD_VERSION_API } from 'repositories/routes';
import { DELETED, PUBLICATION_TYPE_DATA_COLLECTION } from '../config/general';

/**
 * Load publication
 * @param {object}
 * @returns {action}
 */
export function loadRecordToView(pid, isEdit = false) {
    return dispatch => {
        dispatch({ type: actions.VIEW_RECORD_LOADING });

        return get(EXISTING_RECORD_API({ pid: pid.replace('uq:', 'UQ:'), isEdit }))
            .then(response => {
                // if it has been deleted and it's a data collection, handle it as a deleted record
                // this is for cached api responses sitting on public S3 bucket
                if (
                    response.data.rek_status === DELETED &&
                    response.data.rek_display_type === PUBLICATION_TYPE_DATA_COLLECTION
                ) {
                    const error = new Error();
                    error.data = response.data;
                    error.status = 410;
                    return Promise.reject(error);
                }

                dispatch({
                    type: actions.VIEW_RECORD_LOADED,
                    payload: response.data,
                });

                return Promise.resolve(response.data);
            })
            .catch(error => {
                if (error.status === 410) {
                    dispatch({
                        type: actions.VIEW_RECORD_DELETED_LOADED,
                        payload: error.data,
                    });
                } else {
                    dispatch({
                        type: actions.VIEW_RECORD_LOAD_FAILED,
                        payload: error,
                    });
                }
            });
    };
}

/**
 * @param data
 * @return {{}}
 */
export const removeShadowSuffixFromTableNames = data => {
    const normalised = {};
    for (const [key, value] of Object.entries(data)) {
        normalised[key.replace('_shadow', '')] = value;
    }
    return normalised;
};

/**
 * @param {object}
 * @returns {action}
 */
// eslint-disable-next-line no-unused-vars
export function loadRecordVersionToView(pid, version) {
    return dispatch => {
        dispatch({ type: actions.VIEW_RECORD_LOADING });
        return get(EXISTING_RECORD_VERSION_API(pid.replace('uq:', 'UQ:'), version.replace('uq:', 'UQ:')))
            .then(response => {
                response.data = removeShadowSuffixFromTableNames(response.data);
                dispatch({
                    type: actions.VIEW_RECORD_LOADED,
                    payload: response.data,
                });
                return Promise.resolve(response.data);
            })
            .catch(error => {
                if (error.status === 410) {
                    dispatch({
                        type: actions.VIEW_RECORD_VERSION_DELETED_LOADED,
                        payload: removeShadowSuffixFromTableNames(error.data),
                    });
                } else {
                    dispatch({
                        type: actions.VIEW_RECORD_LOAD_FAILED,
                        payload: error,
                    });
                }
            });
    };
}

/**
 * Clear record to be viewed
 * @returns {action}
 */
export function clearRecordToView() {
    return dispatch => {
        dispatch({
            type: actions.VIEW_RECORD_CLEAR,
        });
    };
}

export function unlockRecordToView() {
    return {
        type: actions.VIEW_RECORD_UNLOCK,
    };
}

export function loadDetailedHistory(pid) {
    return dispatch => {
        dispatch({ type: actions.DETAILED_HISTORY_LOADING });
        return get(EXISTING_RECORD_HISTORY_API({ pid: pid.replace('uq:', 'UQ:') }))
            .then(response => {
                dispatch({
                    type: actions.DETAILED_HISTORY_LOADING_SUCCESS,
                    payload: response.data,
                });

                return Promise.resolve(response.data);
            })
            .catch(
                /* istanbul ignore next */ error => {
                    /* istanbul ignore next */
                    dispatch({
                        type: actions.DETAILED_HISTORY_LOADING_FAILED,
                        payload: error,
                    });
                },
            );
    };
}
