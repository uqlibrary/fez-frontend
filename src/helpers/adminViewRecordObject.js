import React from 'react';
import Typography from '@mui/material/Typography';
import PropTypes from 'prop-types';
import { parseHtmlToJSX } from 'helpers/general';
import config from 'locale/viewRecord';
import Button from '@mui/material/Button';
export const navigateToEdit = (navigate, pid) => {
    navigate(`/admin/edit/${pid}?tab=authors`);
};

export const formattedString = (type, lookup) => {
    if (!!!type && !!!lookup) return '-';
    return `${type ?? ''}${type && lookup ? ' - ' : ''}${lookup ?? ''}`;
};

export const parseKey = (key, content) => {
    return key.split('.').reduce((prev, curr) => prev && prev[curr], content);
};

export const authorAffiliates = (key, content, navigate, pid, AAProblems) => {
    const hasError = AAProblems.length > 0;
    const authorAffiliate = parseKey(key, content) ?? null;

    if (!!!authorAffiliate || !Array.isArray(authorAffiliate)) {
        return config.viewRecord.adminViewRecordDrawerFields.hasNoAffiliates;
    }
    if (hasError) {
        const Problems =
            AAProblems.length > 0 &&
            AAProblems.map(item => (
                <Typography
                    variant={'body2'}
                    style={{ marginTop: 10, display: 'block' }}
                    component={'span'}
                    key={`affil_error_${item.rek_author_id}`}
                    id={`affil_error_${item.rek_author_id}`}
                    data-testid={`affil_error_${item.rek_author_id}`}
                    aria-label={`Afffiliation error for ${item.rek_author}`}
                >
                    <b>{item.rek_author}</b> {config.viewRecord.adminViewRecordDrawerFields.errorDetail(item)}
                </Typography>
            ));

        const EditButton = (
            <Button
                key={'affil_cal_error_btn'}
                onClick={() => navigateToEdit(navigate, pid)}
                style={{ marginTop: 10, width: '100%' }}
                variant="outlined"
                id="admin-fix-affiliations-button"
                data-testid="btnFixAffiliations"
            >
                {config.viewRecord.adminViewRecordDrawerFields.buttonLabel}
            </Button>
        );
        return [...Problems, EditButton];
    } else {
        return !!authorAffiliate.length > 0
            ? config.viewRecord.adminViewRecordDrawerFields.hasAffiliates
            : config.viewRecord.adminViewRecordDrawerFields.hasNoAffiliates;
    }
};

export const getDefaultDrawerDescriptorObject = () => {
    return (
        (config?.viewRecord?.adminViewRecordDefaultContent?.object && {
            ...config.viewRecord.adminViewRecordDefaultContent.object,
        }) ||
        /* istanbul ignore next */
        undefined
    );
};
export const getDefaultDrawerDescriptorIndex = () => {
    return (
        (config?.viewRecord?.adminViewRecordDefaultContent?.index && {
            ...config.viewRecord.adminViewRecordDefaultContent.index,
        }) ||
        /* istanbul ignore next */
        undefined
    );
};

/* istanbul ignore next */
export const createDefaultDrawerDescriptorObject = (
    locale = {},
    content = [],
    fields = {},
    navigate,
    pid,
    shouldHandleAffiliations = false,
    AAProblems = [],
) => {
    const adminViewRecordDefaultContentObject = getDefaultDrawerDescriptorObject();
    const adminViewRecordDefaultContentIndex = getDefaultDrawerDescriptorIndex();
    if (!adminViewRecordDefaultContentObject || !adminViewRecordDefaultContentIndex) return {};

    // Notes
    const parsedNotesKey = formattedString(parseKey(fields.notes, content));
    // don't ReactHtmlParse notes if they're empty and the above function has returned '-',
    // otherwise ReactHtmlParse will return an array and will cause issues when rendering to screen.
    const parsedNotes = parsedNotesKey !== '-' ? parseHtmlToJSX(parsedNotesKey) : parsedNotesKey;
    adminViewRecordDefaultContentObject.sections[adminViewRecordDefaultContentIndex.notes][0].value = locale.notes;
    adminViewRecordDefaultContentObject.sections[adminViewRecordDefaultContentIndex.notes][1].value = parsedNotes;

    // Authors (Change this for new AA)
    adminViewRecordDefaultContentObject.sections[adminViewRecordDefaultContentIndex.authors][0].value =
        locale.authorAffiliations;
    adminViewRecordDefaultContentObject.sections[adminViewRecordDefaultContentIndex.authors][0].error =
        !!shouldHandleAffiliations && AAProblems.length > 0;
    adminViewRecordDefaultContentObject.sections[
        adminViewRecordDefaultContentIndex.authors
    ][1].value = !!shouldHandleAffiliations
        ? authorAffiliates(fields.authorAffiliates, content, navigate, pid, AAProblems)
        : config.viewRecord.adminViewRecordDrawerFields.affiliatesDoNotApply;
    // WoS
    adminViewRecordDefaultContentObject.sections[adminViewRecordDefaultContentIndex.wos][0].value = locale.wosId;
    adminViewRecordDefaultContentObject.sections[adminViewRecordDefaultContentIndex.wos][1].value = formattedString(
        parseKey(fields.wosId, content) ?? undefined,
    );
    adminViewRecordDefaultContentObject.sections[adminViewRecordDefaultContentIndex.wos][2].value = locale.wosDocType;
    adminViewRecordDefaultContentObject.sections[adminViewRecordDefaultContentIndex.wos][3].value = formattedString(
        parseKey(fields.wosDocType, content) ?? undefined,
        parseKey(fields.wosDocTypeLookup, content) ?? undefined,
    );

    // Scopus
    adminViewRecordDefaultContentObject.sections[adminViewRecordDefaultContentIndex.scopus][0].value = locale.scopusId;
    adminViewRecordDefaultContentObject.sections[adminViewRecordDefaultContentIndex.scopus][1].value = formattedString(
        parseKey(fields.scopusId, content) ?? undefined,
    );
    adminViewRecordDefaultContentObject.sections[adminViewRecordDefaultContentIndex.scopus][2].value =
        locale.scopusDocType;
    adminViewRecordDefaultContentObject.sections[adminViewRecordDefaultContentIndex.scopus][3].value = formattedString(
        parseKey(fields.scopusDocType, content) ?? undefined,
        parseKey(fields.scopusDocTypeLookup, content) ?? undefined,
    );

    // Pubmed
    adminViewRecordDefaultContentObject.sections[adminViewRecordDefaultContentIndex.pubmed][0].value = locale.pubmedId;
    adminViewRecordDefaultContentObject.sections[adminViewRecordDefaultContentIndex.pubmed][1].value = formattedString(
        parseKey(fields.pubMedId, content) ?? undefined,
    );
    adminViewRecordDefaultContentObject.sections[adminViewRecordDefaultContentIndex.pubmed][2].value =
        locale.pubmedCentralId;
    adminViewRecordDefaultContentObject.sections[adminViewRecordDefaultContentIndex.pubmed][3].value = formattedString(
        parseKey(fields.pubMedCentralId, content) ?? undefined,
    );
    adminViewRecordDefaultContentObject.sections[adminViewRecordDefaultContentIndex.pubmed][4].value =
        locale.pubmedDocType;
    adminViewRecordDefaultContentObject.sections[adminViewRecordDefaultContentIndex.pubmed][5].value = formattedString(
        parseKey(fields.pubMedDocType, content) ?? undefined,
        parseKey(fields.pubMedDocTypeLookup, content) ?? undefined,
    );

    return adminViewRecordDefaultContentObject;
};
createDefaultDrawerDescriptorObject.propTypes = {
    locale: PropTypes.object.isRequired,
    content: PropTypes.array.isRequired,
    fields: PropTypes.object.isRequired,
};
