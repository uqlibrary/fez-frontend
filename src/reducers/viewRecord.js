import * as actions from 'actions/actionTypes';

export const initialState = {
    recordToView: null,
    loadingRecordToView: true,
    recordToViewError: null,
    isRecordLocked: false,
    isDeleted: false,
    isDeletedVersion: false,
    isJobCreated: false,
    error: null,
};

const handlers = {
    [actions.VIEW_RECORD_LOADING]: () => ({
        ...initialState,
    }),

    [actions.VIEW_RECORD_LOADED]: (state, action) => {
        return {
            ...initialState,
            loadingRecordToView: false,
            recordToView: action.payload,
            isRecordLocked: !!action.payload.rek_editing_user,
        };
    },

    [actions.VIEW_RECORD_VERSION_DELETED_LOADED]: (state, action) => ({
        ...initialState,
        isDeletedVersion: true,
        loadingRecordToView: false,
        recordToView: action.payload,
        isRecordLocked: !!action.payload.rek_editing_user,
    }),

    [actions.VIEW_RECORD_LOAD_FAILED]: (state, action) => ({
        ...initialState,
        loadingRecordToView: false,
        recordToViewError: action.payload,
    }),

    [actions.VIEW_RECORD_DELETED_LOADED]: (state, action) => ({
        ...initialState,
        loadingRecordToView: false,
        recordToView: action.payload,
        isDeleted: true,
    }),

    [actions.VIEW_RECORD_CLEAR]: () => ({
        ...initialState,
        isRecordLocked: false,
    }),

    [actions.VIEW_RECORD_UNLOCK]: state => ({
        ...state,
        isRecordLocked: false,
    }),

    [actions.ADMIN_UPDATE_WORK_JOB_CREATED]: state => ({
        ...state,
        isJobCreated: true,
    }),

    [actions.ADMIN_UPDATE_WORK_FAILED]: (state, action) => ({
        ...state,
        error: action.payload,
    }),

    [actions.DETAILED_HISTORY_LOADING]: state => ({
        ...state,
        isLoadingDetailedHistory: true,
    }),
    [actions.DETAILED_HISTORY_LOADING_SUCCESS]: (state, action) => ({
        ...state,
        isLoadingDetailedHistory: false,
        recordDetailedHistory: action.payload,
    }),
    [actions.DETAILED_HISTORY_LOADING_FAILED]: /* istanbul ignore next */ state => /* istanbul ignore next */ ({
        ...state,
        isLoadingDetailedHistory: false,
    }),
};

export default function viewRecordReducer(state = { ...initialState }, action) {
    const handler = handlers[action.type];
    if (!handler) {
        return state;
    }
    return handler(state, action);
}
