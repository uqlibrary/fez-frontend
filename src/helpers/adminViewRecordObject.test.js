import { navigateToEdit, authorAffiliates } from './adminViewRecordObject';
import viewRecord from 'locale/viewRecord';

describe('Author Affiliations', () => {
    const mockNavigate = jest.fn();
    it('Should correctly handle navigation', () => {
        const expected = '/admin/edit/UQ:1?tab=authors';
        navigateToEdit(mockNavigate, 'UQ:1');
        expect(mockNavigate).toBeCalledWith(expected);
    });
    it('should calculate affiliations correctly for valid affiliation data', () => {
        const KEY = 'fez_author_affiliation';
        const CONTENT = {
            fez_author_affiliation: [
                {
                    af_id: 478908,
                    af_pid: 'UQ:764e150',
                    af_author_id: 7624839,
                    af_percent_affiliation: 100000,
                    af_org_id: 1248,
                    af_status: 1,
                    fez_author: {
                        aut_id: 7624839,
                        aut_display_name: 'Lancaster, Steve',
                    },
                    fez_org_structure: [
                        {
                            org_id: 1248,
                            org_title: 'Information Systems and Resource Services (University of Queensland Library)',
                        },
                    ],
                },
                {
                    af_id: 478907,
                    af_pid: 'UQ:764e150',
                    af_author_id: 7624847,
                    af_percent_affiliation: 100000,
                    af_org_id: 1248,
                    af_status: 1,
                    fez_author: {
                        aut_id: 7624847,
                        aut_display_name: 'Sibbald, Lee',
                    },
                    fez_org_structure: [
                        {
                            org_id: 1248,
                            org_title: 'Information Systems and Resource Services (University of Queensland Library)',
                        },
                    ],
                },
            ],
        };

        const result = authorAffiliates(KEY, CONTENT, mockNavigate, 'UQ:1', []);
        expect(result).toEqual(viewRecord.viewRecord.adminViewRecordDrawerFields.hasAffiliates);
    });

    it('should calculate affiliations correctly for no affiliate information', () => {
        const KEY = 'fez_author_affiliation';
        const CONTENT = {
            fez_author_affiliation: null,
        };

        const result = authorAffiliates(KEY, CONTENT, mockNavigate, 'UQ:1', []);
        expect(result).toEqual(viewRecord.viewRecord.adminViewRecordDrawerFields.hasNoAffiliates);
    });

    it('should calculate affiliations correctly for invalid affiliation percentile', () => {
        const KEY = 'fez_author_affiliation';
        const PROBLEMS = [
            {
                rek_author_id: 7624839,
                aut_display_name: 'Lancaster, Steve',
                hasOrgAffiliations: true,
                has100pcAffiliations: false,
            },
        ];
        const CONTENT = {
            fez_author_affiliation: [
                {
                    af_id: 478908,
                    af_pid: 'UQ:764e150',
                    af_author_id: 7624839,
                    af_percent_affiliation: 90000,
                    af_org_id: 1248,
                    af_status: 1,
                    fez_author: {
                        aut_id: 7624839,
                        aut_display_name: 'Lancaster, Steve',
                    },
                    fez_org_structure: [
                        {
                            org_id: 1248,
                            org_title: 'Information Systems and Resource Services (University of Queensland Library)',
                        },
                    ],
                },
                {
                    af_id: 478907,
                    af_pid: 'UQ:764e150',
                    af_author_id: 7624847,
                    af_percent_affiliation: 100000,
                    af_org_id: 1248,
                    af_status: 1,
                    fez_author: {
                        aut_id: 7624847,
                        aut_display_name: 'Sibbald, Lee',
                    },
                    fez_org_structure: [
                        {
                            org_id: 1248,
                            org_title: 'Information Systems and Resource Services (University of Queensland Library)',
                        },
                    ],
                },
            ],
        };

        const result = JSON.stringify(authorAffiliates(KEY, CONTENT, mockNavigate, 'UQ:1', PROBLEMS));
        expect(result).toMatchSnapshot();
    });

    it('should calculate affiliations correctly for orphaned org affiliation', () => {
        const KEY = 'fez_author_affiliation';
        const PROBLEMS = [
            {
                rek_author_id: 7624839,
                aut_display_name: 'Lancaster, Steve',
                hasOrgAffiliations: false,
                has100pcAffiliations: true,
            },
        ];
        const CONTENT = {
            fez_author_affiliation: [
                {
                    af_id: 478908,
                    af_pid: 'UQ:764e150',
                    af_author_id: 7624839,
                    af_percent_affiliation: 100000,
                    af_org_id: 1248,
                    af_status: 1,
                    fez_author: {
                        aut_id: 7624839,
                        aut_display_name: 'Lancaster, Steve',
                    },
                },
                {
                    af_id: 478907,
                    af_pid: 'UQ:764e150',
                    af_author_id: 7624847,
                    af_percent_affiliation: 100000,
                    af_org_id: 1248,
                    af_status: 1,
                    fez_author: {
                        aut_id: 7624847,
                        aut_display_name: 'Sibbald, Lee',
                    },
                    fez_org_structure: [
                        {
                            org_id: 1248,
                            org_title: 'Information Systems and Resource Services (University of Queensland Library)',
                        },
                    ],
                },
            ],
        };

        const result = JSON.stringify(authorAffiliates(KEY, CONTENT, mockNavigate, 'UQ:1', PROBLEMS));
        expect(result).toMatchSnapshot();
    });
});
