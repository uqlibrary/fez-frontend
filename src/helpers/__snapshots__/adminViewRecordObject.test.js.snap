// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Author Affiliations should calculate affiliations correctly for invalid affiliation percentile 1`] = `"[{"type":{"propTypes":{}},"key":"affil_error_7624839","ref":null,"props":{"variant":"body2","style":{"marginTop":10,"display":"block"},"component":"span","id":"affil_error_7624839","data-testid":"affil_error_7624839","aria-label":"Afffiliation error for undefined","children":[{"type":"b","key":null,"ref":null,"props":{},"_owner":null,"_store":{}}," "," has orphaned author information"]},"_owner":null,"_store":{}},{"type":{"propTypes":{}},"key":"affil_cal_error_btn","ref":null,"props":{"style":{"marginTop":10,"width":"100%"},"variant":"outlined","id":"admin-fix-affiliations-button","data-testid":"btnFixAffiliations","children":"Edit Affiliations"},"_owner":null,"_store":{}}]"`;

exports[`Author Affiliations should calculate affiliations correctly for orphaned org affiliation 1`] = `"[{"type":{"propTypes":{}},"key":"affil_error_7624839","ref":null,"props":{"variant":"body2","style":{"marginTop":10,"display":"block"},"component":"span","id":"affil_error_7624839","data-testid":"affil_error_7624839","aria-label":"Afffiliation error for undefined","children":[{"type":"b","key":null,"ref":null,"props":{},"_owner":null,"_store":{}}," "," has orphaned author information"]},"_owner":null,"_store":{}},{"type":{"propTypes":{}},"key":"affil_cal_error_btn","ref":null,"props":{"style":{"marginTop":10,"width":"100%"},"variant":"outlined","id":"admin-fix-affiliations-button","data-testid":"btnFixAffiliations","children":"Edit Affiliations"},"_owner":null,"_store":{}}]"`;
